### PR TITLE
feat: improve Zenith Passage

### PR DESCRIPTION
**Pull request by anna-carroll**
- [x] allow multiple rollups within HostPassage
- [x] enter with ERC20s & ETH
- [x] exit with Host ERC20s & ETH

### CODE REVIEW

File: src/Passage.sol [hunk 1]
```diff
@@ -34,28 +37,46 @@ contract HostPassage {
...
+        emit Enter(rollupChainId, address(0), rollupRecipient, msg.value);
+    }
+
+    /// @notice Allows ERC20s to enter the rollup.
+    /// @dev Permanently burns the token amount by locking it in this contract.
+    /// @param rollupChainId - The rollup chain to enter.
+    /// @param rollupRecipient - The recipient of the Ether on the rollup.
+    /// @param token - The address of the ERC20 token on the Host.
+    /// @param amount - The amount of the ERC20 token to transfer to the rollup.
+    /// @custom:emits Enter indicatig the amount of tokens to mint on the rollup & its recipient.
```
Comment (dylanlott) [inline src/Passage.sol:76]:
```suggestion
    /// @custom:emits Enter indicating the amount of tokens to mint on the rollup & its recipient.
```

File: src/Passage.sol [hunk 2]
```diff
@@ -72,15 +93,23 @@ contract HostPassage {
...
+            if (block.timestamp > orders[i].deadline) revert OrderExpired();
+            // transfer value
+            if (orders[i].token == address(0)) {
+                // transfer native Ether to the recipient
+                payable(orders[i].recipient).transfer(orders[i].amount);
+                // NOTE: this will underflow if sender attempts to transfer more Ether than they sent to the contract
+                etherTransferred -= orders[i].amount;
+            } else {
+                // transfer tokens to the recipient
+                IERC20(orders[i].token).transferFrom(msg.sender, orders[i].recipient, orders[i].amount);
```
Comment (dylanlott) [inline src/Passage.sol:109]: 🙌

**Review by dylanlott — APPROVED** [review]
Looks good to me 🔥 ERC 20 support is a nice win!

File: src/Passage.sol [hunk 3]
```diff
@@ -72,15 +93,23 @@ contract HostPassage {
     ///      the Builder may transfer the cumulative tokenOut to the user in a single ExitFilled event.
     ///      The rollup STF will apply the user's exit transactions on the rollup up to the point that sum(tokenOut) is lte the ExitFilled amount.
     /// TODO: add option to fulfill ExitOrders with native ETH? or is it sufficient to only allow users to exit via WETH?
-    function fulfillExits(ExitOrder[] calldata orders) external {
+    function fulfillExits(ExitOrder[] calldata orders) external payable {
```
Comment (prestwich) [inline src/Passage.sol:96]: why do we perform any checks at all here? given the purpose is for builder to transfer funds, we don't really need to safety check them. they are capable of doing that for themselves.

```
struct Fullfillment {
    // generated somewhere in rollup consensus, must commit to chain ID
    uint256 orderId; 

    address token;
    address recipient;
    uint256 amount;
}
```

Comment (prestwich) [reply src/Passage.sol:96]: basically i think at least the Deadline check is unnecessary, as that part will be performed in the RU consensus

Comment (anna-carroll) [reply src/Passage.sol:96]: > why do we perform any checks at all here? given the purpose is for builder to transfer funds, we don't really need to safety check them. they are capable of doing that for themselves.

i don't understand what you mean 👀 the thinking behind this fn is to emit events that the RU node can use to easily see that funds have been transferred in accordance with the intent expressed on L2. we perform checks to ensure that said events actually correspond to funds transfers

Comment (prestwich) [reply src/Passage.sol:96]: Zenith doesnt need to perform any check that there is a corresponding RU tx that the builder is trying to fill

Comment (prestwich) [reply src/Passage.sol:96]: (and indeed can't effectively perform any check like that)

Comment (prestwich) [reply src/Passage.sol:96]: the builder is capable of making that check outside this system, if they want to. we dont need to make it mandatory for them by putting it in-system

Comment (prestwich) [reply src/Passage.sol:96]: > ah do you just mean the Builder adds a block deadline

yes they can define a confirmby :)

Comment (anna-carroll) [reply src/Passage.sol:97]: i removed the deadline check: https://github.com/init4tech/zenith/pull/11/commits/53982a5b09645a035536e889585ab67e04797032

Comment (anna-carroll) [reply src/Passage.sol:97]: not sure about `orderId` - how would you propose handling that in the RU node?

File: src/Passage.sol [hunk 4]
```diff
@@ -72,15 +93,23 @@ contract HostPassage {
...
+        uint256 etherTransferred = msg.value;
```
[the remark this answers is not in the document]
Comment (anna-carroll) [reply src/Passage.sol]: https://github.com/init4tech/zenith/pull/11/commits/dd56aa065d96b41305d68d5457b20ea037e87353

File: src/Passage.sol (hunk 1 above)
Comment (anna-carroll) [reply src/Passage.sol]: https://github.com/init4tech/zenith/pull/11/commits/c5f4dfa712440f048f328a18f7e2dbf4d61bcef9

File: src/Passage.sol [hunk 5]
```diff
@@ -154,7 +182,7 @@ contract RollupPassage {
     /// @param amountOutMinimum_H - The minimum amount of tokenOut_H the user expects to receive on host.
     /// @custom:reverts Expired if the deadline has passed.
     /// @custom:emits Exit if the exit transaction succeeds.
-    function submitExit(address tokenOut_H, address recipient_H, uint256 deadline, uint256 amountOutMinimum_H)
+    function submitEthExit(address tokenOut_H, address recipient_H, uint256 deadline, uint256 amountOutMinimum_H)
```
Comment (prestwich) [inline src/Passage.sol:185]: i;m not sure i understand the expected exit flow here. when is this function called?

**Review by prestwich — APPROVED** [review]

**Pull request merged**